File: python/37.py
from typing import List
from itertools import repeat
import numpy as np

class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        n, m = len(board), len(board[0])
        
        rows = np.array([[False for _ in range(9)]for _ in range(9)])
        cols = np.array([[False for _ in range(9)]for _ in range(9)])
        cells = np.array([[[False for _ in range(9)]for _ in range(9)] for _ in range(3)])

        # rows =  [  dict(zip(range(1, 10), repeat(False, 9))) for i in range(n)]  # 记录行
        # cols =  [  dict(zip(range(1, 10), repeat(False, 9))) for i in range(m)]  # 记录列
        # cells = [[ dict(zip(range(1, 10), repeat(False, 9))) for i in range(n//3)] for j in range(m//3)]  # 记录cells
        
        cnt = 0
        
        for i in range(n):
            for j in range(m):
                if board[i][j] == '.':
                    cnt += 1
                    continue
                idx = ord(board[i][j]) - ord('1')
                
                rows[i][idx] = True
                cols[j][idx] = True
        
        def getPossibleStatus(i, j):
            """
            返回可能位置的一个bool向量
            """
            return ~(rows[i] | rows[j] | cells[i//3][j//3])
        
        def getNext():  # 动态获得下一个
            """
            找到范围最小的点
            """
            minCnt = 10
            res = (0, 0)
            for i in range(9):
                for j in range(9):
                    if board[i][j] != '.': continue
                    cur = getPossibleStatus(i, j)
                    
                    if np.sum(cur) < minCnt:
                       res = (i, j)
                       minCnt = np.sum(cur)
            return res
        
        def fillNum(x:int, y:int, n:int, flag:bool) -> None:
            rows[x][n] = flag
            cols[y][n] = flag
            cells[x//3][y//3][n] = flag
        
        def dfs(cnt: int) -> bool:
            if cnt == 0: return True
            
            next_pos = getNext()
            bit_vec = getPossibleStatus(next_pos[0], next_pos[1])
            for idx in range(9):
                if not bit_vec[idx]: continue
                fillNum(next_pos[0], next_pos[1], idx, True)
                board[next_pos[0]][next_pos[1]] = chr(idx + ord('1'))
                if dfs(cnt-1): return True
                board[next_pos[0]][next_pos[1]] = '.'
                fillNum(next_pos[0], next_pos[1], idx, False)
                               
            return False
    
        dfs(cnt)
        print(board)
        
            
s = Solution()
board = eval("""[["5","3",".",".","7",".",".",".","."],["6",".",".","1","9","5",".",".","."],[".","9","8",".",".",".",".","6","."],["8",".",".",".","6",".",".",".","3"],["4",".",".","8",".","3",".",".","1"],["7",".",".",".","2",".",".",".","6"],[".","6",".",".",".",".","2","8","."],[".",".",".","4","1","9",".",".","5"],[".",".",".",".","8",".",".","7","9"]]""")
s.solveSudoku(board=board)

File: cpp/37.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /**
     *  Leetcode题解【解数独】回溯 + 状态压缩（使用 bitset）
    */
    void solveSudoku(vector<vector<char>>& board) {
        rows = vector<bitset<9>>(9, bitset<9>());  // 行的标记, 一行9个
        cols = vector<bitset<9>>(9, bitset<9>());  // 列的标记
        cells = vector<vector<bitset<9>>>(3, vector<bitset<9>>(3, bitset<9>()));  // cell 部分， 记录数字
        
        int cnt = 0;  // 计数单元，记录有多少个待填的数
        int n = board.size(), m = board[0].size();
        for (int i=0; i < n; ++i) {
            for (int j=0; j < m; ++j) {
                cnt += (board[i][j] == '.');
                if (board[i][j] == '.') continue;
                int n = board[i][j] - '1';  // 1-9 => 0-8
                rows[i] |= (1<<n);  // 标记已经选择的数
                cols[j] |= (1<<n);  // 标记已经选择的数
                cells[i/3][j/3] |= (1<<n); // cells中标记已经选择的数
            }
        }
        dfs(board, cnt);   
    }
    bool dfs(vector<vector<char>>& board, int cnt) {
        if (cnt == 0) return true;

        auto next = getNext(board);
        auto bits = getPossibleStatus(next[0], next[1]);
        for (int n = 0; n < bits.size(); ++n) {
            if (!bits.test(n)) continue;            // continue
            fillNum(next[0], next[1], n, true);     // 标记
            board[next[0]][next[1]] = n+'1';
            if (dfs(board, cnt - 1)) return true;
            board[next[0]][next[1]] = '.';
            fillNum(next[0], next[1], n, false);    // 恢复
        }
        return false;
    }

    vector<int> getNext(vector<vector<char>>& board) {
        vector<int> ret;
        int minCnt = 10;  // 元素个数
        int n = board.size(), m = board[0].size();
        for (int i=0; i < n; ++i) {
            for (int j=0; j < m; ++j) {
                if (board[i][j] != '.') continue;
                auto cur = getPossibleStatus(i, j);  // 还没填的位置，获取其 可以填啥
                // 找到最小范围的位置
                if (cur.count() >= minCnt) continue; // 如果没有
                ret = {i, j};
                minCnt = cur.count(); // 或者set位置
            }
        }
        return ret;
    }

    bitset<9> getPossibleStatus(int x, int y) {
        return ~(rows[x] | cols[y] | cells[x/3][y/3]);  // get 未标记的
    }

    void fillNum(int x, int y, int n, bool fillFlag) {
        rows[x][n] = (fillFlag)? 1:0;
        cols[y][n] = (fillFlag)? 1:0;
        cells[x/3][y/3][n] = (fillFlag)? 1:0;
    }

private:
    vector<bitset<9>> rows;
    vector<bitset<9>> cols;
    vector<vector<bitset<9>>> cells;
};